cellprofiler/modules/savecroppedobjects.py

```diff
@@ -126,67 +210,78 @@ def run(self, workspace):
                 )
 
-                skimage.io.imsave(
-                    filename, skimage.img_as_ubyte(mask), check_contrast=False
+            elif not self.nested_save.value:
+                filename = os.path.join(
+                    directory, "{}{}_{}".format(input_filename + "_" if self.use_filename else "", self.objects_name.value, label)
                 )
 
-            elif self.file_format.value == O_TIFF_8:
-                filename = os.path.join(
-                    directory, "{}_{}.{}".format(self.objects_name.value, label, "tiff")
+            if self.file_format.value == O_PNG:
+                save_filename = filename + ".{}".format(O_PNG)
+
+                skimage.io.imsave(
+                    save_filename, 
+                    skimage.img_as_ubyte(mask), 
+                    check_contrast=False
                 )
 
+            elif self.file_format.value == O_TIFF_8:
+                save_filename = filename +".{}".format("tiff")
+                
                 skimage.io.imsave(
-                    filename,
+                    save_filename,
                     skimage.img_as_ubyte(mask),
                     compress=6,
                     check_contrast=False,
                 )
 
             elif self.file_format.value == O_TIFF_16:
-                filename = os.path.join(
-                    directory, "{}_{}.{}".format(self.objects_name.value, label, "tiff")
-                )
-
+                save_filename = filename + ".{}".format("tiff")
+                
                 skimage.io.imsave(
-                    filename,
+                    save_filename,
                     skimage.img_as_uint(mask),
                     compress=6,
                     check_contrast=False,
                 )
 
-            filenames.append(filename)
+            filenames.append(save_filename)
 
         if self.show_window:
             workspace.display_data.filenames = filenames
 
-    def settings(self):
-        settings = [
-            self.objects_name,
-            self.directory,
-            self.file_format,
-            self.export_option,
-            self.image_name,
-        ]
-
-        return settings
-
-    def visible_settings(self):
-        result = [
-            self.export_option,
-            self.objects_name,
-            self.directory,
-            self.file_format,
-        ]
-
-        if self.export_option.value == SAVE_PER_OBJECT:
-            result += [self.image_name]
-
-        return result
+    def upgrade_settings(self, setting_values, variable_revision_number, module_name):
+        if variable_revision_number == 1:
+            # Old order: 
+            # [objects_name, directory, file_format]
+            # New order:
+            # [objects_name, directory, file_format, export_option, image_name]
+            setting_values = (
+                [setting_values[:3] + [SAVE_PER_OBJECT, "Image"]]
```

Comment:
Can you check the data structure here and below? I think you're starting from a single un-nested list and ending in a nested list.

Reply:
Actually, below looks fine, I think it's just here

Reply:
Ah, I didn't spot that. Thank you